lib/BackgroundJob/MailNotificationSender.php

```diff
@@ -0,0 +1,192 @@
+		$this->logger = $logger ? $logger : \OC::$server->getLogger();
+		$this->mailer = $mailer ? $mailer : \OC::$server->getMailer();
+		$this->config = $config ? $config : \OC::$server->getConfig();
+
+		$optionsStorage = new OptionsStorage($this->config);
+		$notificationMailer = new NotificationMailer($this->notificationManager, $this->mailer, $optionsStorage);
+
+		$this->mailerAdapter = $mailerAdapter ? $mailerAdapter :
+			new NotificationMailerAdapter($notificationMailer, $this->userManager, $this->logger, $this->urlGenerator);
+	}
+
+	/**
+	 * @param $notificationURL
+	 * @return string
+	 */
+	public function fixURL($notificationURL) {
+		$url = $this->request->getServerProtocol() . '://' . $this->request->getServerHost();
+		$partURL = \explode($url, $notificationURL)[1];
+		$webRoot = $this->getArgument()['webroot'];
+		if (\strpos($partURL, $webRoot) === false) {
+			$notificationURL = $url . $this->getArgument()['webroot'] . $partURL;
+		}
+		return $notificationURL;
+	}
+
+	/**
+	 * @param $shareFullId
+	 */
+	public function sendNotify($shareFullId) {
+		$notificationList = [];
+		$maxCountSendNotification = 100;
+
+		//Check if its an internal share or not
+		try {
+			$share = $this->shareManager->getShareById($shareFullId);
+		} catch (ShareNotFound $e) {
+			return null;
+		}
+
+		$users = [];
+		if ($share->getShareType() === \OCP\Share::SHARE_TYPE_GROUP) {
+			//Notify all the group members
+			$group = $this->groupManager->get($share->getSharedWith());
+			$users = $group->getUsers();
+		} elseif ($share->getShareType() === \OCP\Share::SHARE_TYPE_USER) {
+			$users[] = $this->userManager->get($share->getSharedWith());
+		}
+
+		foreach ($users as $user) {
+			if ($user->getEMailAddress() === null) {
+				\OC::$server->getLogger()->warning("Mail notification can not be sent to " . $user->getUID() . " for share " . $share->getName() .  " as email for user is not set");
+				continue;
+			}
+			$notification = $this->notificationManager->createNotification();
+			$notification->setApp('files_sharing')
+				->setUser($user->getUID())
+				->setDateTime(new \DateTime())
+				->setObject('local_share', $shareFullId);
+
+			$fileLink = $this->urlGenerator->linkToRouteAbsolute('files.viewcontroller.showFile', ['fileId' => $share->getNode()->getId()]);
+
+			$fileLink = $this->fixURL($fileLink);
+			$notification->setLink($fileLink);
+
+			$notification->setSubject('local_share', [$share->getShareOwner(), $share->getSharedBy(), $share->getNode()->getName()]);
+			$notification->setMessage('local_share', [$share->getShareOwner(), $share->getSharedBy(), $share->getNode()->getName()]);
+
+			//Adding action because sendMail requires an action for the notificaiton
+			$acceptAction = $notification->createAction();
+			$acceptAction->setLabel('email');
+			$acceptAction->setLink($fileLink, 'POST');
+			$notification->addAction($acceptAction);
+
+			if (\count($notificationList) < $maxCountSendNotification) {
```

Comment:
does it make a difference to send the emails like this? It took a while to understand what you're doing here. Sending the emails one by one would be easier, so unless there is a real advantage I'd prefer to go through the easy route.